SCClassLibrary/Common/Control/BelaServerOptions.sc

```diff
@@ -0,0 +1,55 @@
+BelaServerOptions {
+
+	classvar defaultValues;
+
+	*initDefaults {
+		defaultValues = IdentityDictionary.newFrom(
+			(
+				numAnalogInChannels: 2,
+				numAnalogOutChannels: 2,
+				numDigitalChannels: 16,
+				headphoneLevel: -6,
+				pgaGainLeft: 10,
+				pgaGainRight: 10,
+				speakerMuted: 0,
+				dacLevel: 0,
+				adcLevel: 0,
+				numMultiplexChannels: 0,
+				belaPRU: 1,
+				belaMaxScopeChannels: 0
+			)
+		);
+		^defaultValues;
+	}
+
+	*addBelaOptions { |serverOptions|
```

Comment:
is the conversation here relevant? https://github.com/BelaPlatform/supercollider/pull/85#issuecomment-736928916

Reply:
@brian had suggested to entirely do without a dedicated BelaServerOptions class, but he conceded that in a separate file, such a class is a good thing. Looking at the class, I opt for subclassing `ServerOptions`, and adding instance variables for the parameters.

`asOptionsString` should be concistent with `ServerOptions:asOptionsString`, taking the port as argument (or do you see a reason why this wouldn't be appropriate?).

Here is an untested suggestion:

```supercollider
BelaServerOptions : ServerOptions {
	
	var <>numAnalogInChannels;
	var <>numAnalogOutChannels;
	var <>numDigitalChannels;
	var <>headphoneLevel;
	var <>pgaGainLeft;
	var <>pgaGainRight;
	var <>speakerMuted;
	var <>dacLevel;
	var <>adcLevel;
	var <>numMultiplexChannels;
	var <>belaPRU;
	var <>belaMaxScopeChannels;
	var <>numMultiplexChannels;

	classvar defaultValues;
	
	*initClass {
		defaultValues = super.defaultValues.copy.putAll(
			(
				numAnalogInChannels: 2,
				numAnalogOutChannels: 2,
				numDigitalChannels: 16,
				headphoneLevel: -6,
				pgaGainLeft: 10,
				pgaGainRight: 10,
				speakerMuted: 0,
				dacLevel: 0,
				adcLevel: 0,
				numMultiplexChannels: 0,
				belaPRU: 1,
				belaMaxScopeChannels: 0
			)
		)
	}
	
	device {
		// do be done …
	}

	device_ { |dev|
		// do be done … // maybe not allow?
	}
	
	asOptionsString { | port = 57110 |
		var o = super.asOptionsString(port);
		o = o ++ " -J " ++ numAnalogInChannels;
		o = o ++ " -K " ++ numAnalogOutChannels;
		o = o ++ " -G " ++ numDigitalChannels;
		o = o ++ " -Q " ++ headphoneLevel;
		o = o ++ " -X " ++ pgaGainLeft;
		o = o ++ " -Y " ++ pgaGainRight;
		o = o ++ " -A " ++ speakerMuted;
		o = o ++ " -x " ++ dacLevel;
		o = o ++ " -y " ++ adcLevel;
		o = o ++ " -g " ++ numMultiplexChannels;
		o = o ++ " -T " ++ belaPRU;
		o = o ++ " -E " ++ belaMaxScopeChannels;
		^o
	}
}
```

Reply:
Yes, I'm up for making BelaServerOptions a subclass of ServerOptions. But I think it would be nice to retain its automatic management: I would like Server to choose BelaServerOptions on `init` or `*fromName` if `Platform.hasBelaSupport`, otherwise all code meant to be run on a Bela  should explicitly create Servers with BelaServerOptions. What do you think? Better to leave it explicit for the user to set?

Reply:
Hm I see that you want to make Bela supported transparently.

Not so sure, but here would be a first approach in `Server`:

```supercollider
	var <>defaultOptions;

	*initClass {
		<SNIP>
		defaultOptions = if(Platform.isBela) { BelaServerOptions.new } { ServerOptions.new };
		<SNIP>
	}

	*fromName { |name|
		^Server.named[name] ?? {
			Server(name, NetAddr.new("127.0.0.1", 57110)) // <== removing the unnecessary ServerOptions.new
		}
	}

	init { |argName, argAddr, argOptions, argClientID|
		this.addr = argAddr;
		options = argOptions ?? { this.defaultOptions };
		<SNIP>
	}

	
```

Reply:
All in all, it indicates that the `ServerOptions`'s design is not very good. But for now the best is to be explicit (your solution was smart, but better be blunt).

Reply:
before you go ahead and subclass serveroptions you should make sure that is actually supported by the tight c++ level integration

Reply:
also my pronouns are they/them @telephon

Reply:
@brianlheim thanks for reminding! You already mentioned c++ integration in a comment on Bela's repo, sorry I forgot! So, I'm now looking for it, and the only mention of "Server" I found is in `LangPrimSource/OSCData.cpp`, in function `prBootInProcessServer` which is used when booting the internal server. And it turns out that Bela has never added its options there! Would it be important to do it? If this never came out as a problem, isn't it probably a sign that the internal server is not really used, at least by Bela users? Anyway I would feel bad in leaving it so prone to break!

So, to me, the way to go now looks like subclassing ServerOptions, updating also `prBootInProcessServer`, and requiring `Server.default = Server(BelaServerOptions())` for patches which are supposed to boot a Bela server, adding it to all relevant examples  (which by the way will make it clearer if one example is supposed to be run on a Bela or on a remote host).
If we agree, I can make a patch, test it, and post it here.

@giuliomoro just as a reminder: if this will be the way we implement it, we'll need to change Bela's default SC project too, and especially mention it in docs and release notes.

Reply:
i think @jamshark70 has said before the internal server is unused/broken for a while now? not sure

after making the subclass i think it would still be convenient to have BelaServerOptions be the default options type when working on Bela, and require explicit instantiation on other platforms. unless people think that would be too confusing

Reply:
Just made a PR to this PR's branch: https://github.com/BelaPlatform/supercollider/pull/86

Among other fixes, I implemented BelaServerOptions as a subclass of ServerOptions, and let Server choose BelaServerOptions by default if `Platform.hasBelaSupport`. I'm more and more for this feature, because otherwise all Bela code should start by creating a Server, giving it a name and an address, then BelaServerOptions, or otherwise server won't boot. I put the ServerOptions selection logic as a case, maybe in the future we could have different platform-specific ServerOptions subclasses.
But again, we can decide that we need to be blunt and explicit, and I would remove the automatic selection bit. 

It's important to reach a consensus about this before this PR is merged, since it could eventually require changes in SuperCollider code written for Bela.

Reply:
@elgiano 
> So, I'm now looking for it, and the only mention of "Server" I found is in LangPrimSource/OSCData.cpp, in function prBootInProcessServer which is used when booting the internal server. And it turns out that Bela has never added its options there! Would it be important to do it? If this never came out as a problem, isn't it probably a sign that the internal server is not really used, at least by Bela users? Anyway I would feel bad in leaving it so prone to break!

On Bela you definately don't want to run the server in an internal process of the language. You want it in the Bela audio framework.
So it is not needed there.

Reply:
Here's @elgiano's commit where they subclass `ServerOptions`. https://github.com/BelaPlatform/supercollider/commit/704bce7fe175cd7a24a4be0cc7caabbe0c086910

Does this look OK ?